### PR TITLE
[SPARK-28195][SQL] Fix CheckAnalysis not working for InsertIntoDataSourceDirCommand and report misleading error message

### PR DESCRIPTION
**Pull request by liupc**
## What changes were proposed in this pull request?

This PR will try to fix the issue that the sub plan of InsertDataSourceDirCommand is not being checked for analysis and sometimes will report misleading error message.
An example sql is like 
`insert overwrite directory '/path' using parquet select * from table1`

When "table1" does not exists, we will finally got a misleading error message:
```
Caused by: org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to dataType on unresolved object, tree: 'kr.objective_id
at org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute.dataType(unresolved.scala:105)
at org.apache.spark.sql.types.StructType$$anonfun$fromAttributes$1.apply(StructType.scala:440)
at org.apache.spark.sql.types.StructType$$anonfun$fromAttributes$1.apply(StructType.scala:440)
at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
at scala.collection.immutable.List.foreach(List.scala:381)
at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
at scala.collection.immutable.List.map(List.scala:285)
at org.apache.spark.sql.types.StructType$.fromAttributes(StructType.scala:440)
at org.apache.spark.sql.catalyst.plans.QueryPlan.schema$lzycompute(QueryPlan.scala:159)
at org.apache.spark.sql.catalyst.plans.QueryPlan.schema(QueryPlan.scala:159)
at org.apache.spark.sql.execution.datasources.DataSource.planForWriting(DataSource.scala:544)
at org.apache.spark.sql.execution.command.InsertIntoDataSourceDirCommand.run(InsertIntoDataSourceDirCommand.scala:70)
at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:79)
at org.apache.spark.sql.execution.adaptive.QueryStage.executeCollect(QueryStage.scala:246)
at org.apache.spark.sql.Dataset$$anonfun$6.apply(Dataset.scala:190)
at org.apache.spark.sql.Dataset$$anonfun$6.apply(Dataset.scala:190)
at org.apache.spark.sql.Dataset$$anonfun$52.apply(Dataset.scala:3277)
at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:77)
at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3276)
at org.apache.spark.sql.Dataset.&lt;init&gt;(Dataset.scala:190)
at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:75)
at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:642)
at org.apache.spark.sql.SQLContext.sql(SQLContext.scala:694)
at org.apache.spark.sql.hive.thriftserver.SparkExecuteStatementOperation.org$apache$spark$sql$hive$thriftserver$SparkExecuteStatementOperation$$execute(SparkExecuteStatementOperation.scala:277)
... 11 more
```

## How was this patch tested?

exist UT(AnalysisSuite)

### CODE REVIEW

**Comment by maropu** [comment]
ok to test

**Comment by maropu** [comment]
This issue only exists in INSERT command? Probably, you'd be better to make the title and description more concrete...

**Comment by liupc** [comment]
I checked the command implementations, I'm not very sure, but it seems only affect the INSERT command, thank you! @maropu

**Comment by maropu** [comment]
Also, can you add some tests? This pr will change exception messages, right?

**Comment by liupc** [comment]
OK, I will add some test later.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/CommandAnalysisSuite.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      super.checkAnalysis(planToCheck)
+    }
+  }
+
+  test("SPARK-28195: checkAnalysis should work for InsertIntoDataSourceDirCommand") {
+    val query = "insert overwrite directory '/path' using parquet select * from table1"
+    val exception = intercept[AnalysisException](
+      analyzer.executeAndCheck(parser.parsePlan(query), new QueryPlanningTracker))
+    assert(exception.getMessage.contains("Table or view not found: table1"))
+  }
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/CommandAnalysisSuite.scala]: Can we move this into `InsertSuite`?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala [hunk 2]
```diff
@@ -188,6 +189,15 @@ abstract class BaseSessionStateBuilder(
...
 
     override protected def lookupCatalog(name: String): CatalogPlugin = session.catalog(name)
+
+    override def checkAnalysis(plan: LogicalPlan): Unit = {
+      // We should check it's innerChildren for InsertIntoDataSourceDirCommand
+      val planToCheck = plan match {
+        case e: InsertIntoDataSourceDirCommand => e.query
+        case _ => plan
+      }
+      super.checkAnalysis(planToCheck)
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala]: Can we resolve this issue in the `InsertIntoDataSourceDirCommand` side?

Comment (liupc) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala]: @maropu  Yes, I agree that resolving this issue inside `InsertIntoDataSourceDirCommand` is better, but as I have already questioned is the jira: https://issues.apache.org/jira/browse/SPARK-28195, I'm not sure whether there are some special consideration for making the children of `InsertIntoDataSourceDirCommand` to empty and use innerChildren instead.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/CommandAnalysisSuite.scala (hunk 1 above)
Comment (liupc) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/CommandAnalysisSuite.scala]: It's OK

File: sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala (hunk 2 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala]: You meant this issue only happened when `spark.sql.runSQLOnFiles`=true? What if `spark.sql.runSQLOnFiles`=false?

Comment (liupc) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala]: @maropu This issue affect more aspect, what I point out in the jira is just for the case "table or view not found", actually, this issue may cause many other problems.
The root cause is that the the children of InsertIntoDataSourceDirCommand is empty, thus many analysis rules after InsertIntoDataSourceDirCommand being inserted(In DataSourceAnalysis rule) may not be effective.
I think we can fix it better inside InsertIntoDataSourceDirCommand, but I should first make it clear that why we set it's children to empty, but use innerChildren instead? Is there any PR or issue for that?

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala]: I think you should investigate it by yourself.  I don't like the current approach too. Appeartly issue is minor but the fix is pretty invasive. I doesn't need to touch SessionStateBuilder side at all.

**Comment by HyukjinKwon** [comment]
Closing this due to inactivity from its author.

**Pull request closed without merging**